arbor/communication/mpi_context.cpp

```diff
@@ -24,9 +26,18 @@ struct mpi_context_impl {
     int rank_;
     MPI_Comm comm_;
 
-    explicit mpi_context_impl(MPI_Comm comm): comm_(comm) {
+    explicit mpi_context_impl(MPI_Comm comm, bool bind=false): comm_(comm) {
         size_ = mpi::size(comm_);
         rank_ = mpi::rank(comm_);
+        if (bind) {
+            MPI_Comm local;
+            MPI_Comm_split_type(comm, MPI_COMM_TYPE_SHARED, rank_, MPI_INFO_NULL, &local);
+            auto mpi_guard = util::on_scope_exit([&] { MPI_Comm_free(&local); });
+            int local_rank = -1, local_size = -1;
+            MPI_Comm_rank(local, &local_rank);
+            MPI_Comm_size(local, &local_size);
```

Comment:
Two points here:
- Maybe use `MPI_Comm_free(&local)` to clean up the communicator afterwards
- Why not use the mpi functions from `mpi.hpp`? These would include error checks.

Reply:
- See l 35. It's also exception safe ;)
- Good point.

Reply:
C++ bindings are more or less deprecated I believe

Reply:
@AdhocMan talks about our internal wrappers.

Reply:
And yes, MPIC++ as been deprecated since (pre?) 3.0, I think.